accounts-db/src/accounts_index/in_mem_accounts_index.rs

```diff
@@ -1075,118 +1081,109 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> InMemAccountsIndex<T,
-                        loop {
-                            let disk_resize =
-                                disk.try_write(&key, (&disk_entry, disk_ref_count.into()));
-                            match disk_resize {
-                                Ok(_) => {
-                                    // successfully written to disk
-                                    flush_stats.flush_entries_updated_on_disk += 1;
-                                    break;
-                                }
-                                Err(err) => {
-                                    // disk needs to resize. This item did not get written. Resize and try again.
-                                    let m = Measure::start("flush_grow");
-                                    disk.grow(err);
-                                    flush_stats.flush_grow_us += m.end_as_us();
-                                }
+                        // since we know slot_list.len() == 1, we can create a stack-allocated array for single element
+                        let (slot, info) = slot_list[0];
+                        let disk_entry = [(slot, info.into())];
+
+                        (disk_entry, ref_count)
+                    };
+
+                    flush_stats.flush_read_lock_us += lock_measure.end_as_us();
+
+                    // Now write to disk WITHOUT holding any locks
+                    // may have to loop if disk has to grow and we have to retry the write
+                    loop {
+                        let disk_resize =
+                            disk.try_write(&key, (&disk_entry, disk_ref_count.into()));
+                        match disk_resize {
+                            Ok(_) => {
+                                // successfully written to disk
+                                flush_stats.flush_entries_updated_on_disk += 1;
+                                break;
+                            }
+                            Err(err) => {
+                                // disk needs to resize. This item did not get written. Resize and try again.
+                                let m = Measure::start("flush_grow");
+                                disk.grow(err);
+                                flush_stats.flush_grow_us += m.end_as_us();
                             }
                         }
-
-                        Some(key)
                     }
+
+                    Some(key)
                 })
                 .collect();
 
             flush_stats.flush_update_us = flush_update_measure.end_as_us();
             flush_stats.update_to_stats(self.stats());
 
             let m = Measure::start("flush_evict");
-            self.evict_from_cache(evictions_age, current_age, ages_flushing_now);
+            self.evict_from_cache(&flushed_keys_to_evict, current_age, ages_flushing_now);
+            self.evict_from_cache(&candidates_to_evict.0, current_age, ages_flushing_now);
```

Comment:
This is to preserve existing behavior. A follow-up PR will be to not evict `flushed_keys_to_evict`.